File: app/src/main/java/org/eztarget/papeler/SetWallpaperActivity.java
package org.eztarget.papeler;

import android.app.Activity;
import android.app.WallpaperManager;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

/**
 * Created by michelsievers on 23/01/2017.
 */

public class SetWallpaperActivity extends Activity {

    private boolean mDidSetWallpaper = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (mDidSetWallpaper) {
            finish();
        } else {
            setWallpaper();
        }
    }

    private void setWallpaper() {

        final Intent setLivePapelerIntent;
        setLivePapelerIntent = new Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER);
        setLivePapelerIntent.putExtra(
                WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT,
                new ComponentName(this, WallpapelerService.class)
        );

        startActivity(setLivePapelerIntent);
        mDidSetWallpaper = true;
    }

    public void onClick(View view) {
    }
}

File: app/src/main/java/org/eztarget/papeler/SetWallpaperActivity.kt
package org.eztarget.papeler

import android.app.Activity
import android.app.WallpaperManager
import android.content.ActivityNotFoundException
import android.content.ComponentName
import android.content.Intent
import android.os.Bundle
import android.view.View

class SetWallpaperActivity : Activity() {

    private var didSetWallpaper = false
    var previewWallpaperIntent = Intent(WallpaperManager.ACTION_CHANGE_LIVE_WALLPAPER)
    var selectWallpaperIntent = Intent(WallpaperManager.ACTION_LIVE_WALLPAPER_CHOOSER)

    override fun onResume() {
        super.onResume()

        if (didSetWallpaper) {
            finish()
        } else {
            setWallpaper()
        }
    }

    private fun setWallpaper() {
        previewWallpaperIntent.putExtra(
                WallpaperManager.EXTRA_LIVE_WALLPAPER_COMPONENT,
                ComponentName(this, WallpaperService::class.java)
        )

        try {
            startActivity(previewWallpaperIntent)
        } catch (e: ActivityNotFoundException) {
            startActivity(selectWallpaperIntent)
        }

        didSetWallpaper = true
    }

    fun onClick(view: View) {}
}
